src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/render/ModuleDamageParticles.kt

```diff
@@ -0,0 +1,146 @@
+package net.ccbluex.liquidbounce.features.module.modules.render
+
+import com.mojang.blaze3d.systems.RenderSystem
+import net.ccbluex.liquidbounce.event.events.OverlayRenderEvent
+import net.ccbluex.liquidbounce.event.handler
+import net.ccbluex.liquidbounce.event.repeatable
+import net.ccbluex.liquidbounce.features.module.Category
+import net.ccbluex.liquidbounce.features.module.Module
+import net.ccbluex.liquidbounce.features.module.modules.render.nametags.ModuleNametags
+import net.ccbluex.liquidbounce.render.Fonts
+import net.ccbluex.liquidbounce.render.RenderEnvironment
+import net.ccbluex.liquidbounce.render.engine.Color4b
+import net.ccbluex.liquidbounce.render.engine.font.FontRendererBuffers
+import net.ccbluex.liquidbounce.render.renderEnvironmentForGUI
+import net.ccbluex.liquidbounce.render.withColor
+import net.ccbluex.liquidbounce.utils.client.chat
+import net.ccbluex.liquidbounce.utils.combat.EnemyConfigurable
+import net.ccbluex.liquidbounce.utils.combat.shouldBeShown
+import net.ccbluex.liquidbounce.utils.math.plus
+import net.ccbluex.liquidbounce.utils.render.WorldToScreen
+import net.minecraft.entity.LivingEntity
+import net.minecraft.util.math.Vec3d
+import org.lwjgl.opengl.GL11
+import kotlin.math.PI
+import kotlin.math.absoluteValue
+import kotlin.math.cos
+import kotlin.random.Random
+
+object ModuleDamageParticles : Module("DamageParticles", Category.RENDER) {
+
+    private val duration by int("Duration", 120, 20..360, "ticks")
+    private val scale by float("Scale", 2F, 0.25F..4F)
+    private val maximumDistance by float("MaximumDistance", 32F, 1F..256F)
+
+    private val particles = hashSetOf<Particle>()
+    private val hpMap = hashMapOf<LivingEntity, Float>()
+
+    private val fontRenderer
+        get() = Fonts.DEFAULT_FONT.get()
+
+    override fun disable() {
+        hpMap.clear()
+    }
+
+    val repeatable = repeatable {
+        world.entities.filterIsInstance<LivingEntity>()
+            .filter { it.shouldBeShown() && player.distanceTo(it) <= maximumDistance }
+            .forEach {
+                if (!hpMap.containsKey(it)) {
+                    hpMap[it] = it.health
+                    return@forEach
+                }
+
+                val delta = it.health - hpMap[it]!!
+                if (delta.absoluteValue < 1e-5) return@forEach
+
+                particles.add(
+                    Particle(
+                        String.format("%.1f", delta.absoluteValue),
+                        if (delta < 0) Color4b.RED else Color4b.GREEN,
+                        it.pos + Vec3d(
+                            Random.nextDouble(-0.3, 0.3) + it.velocity.x,
+                            it.height * 0.6,
+                            Random.nextDouble(-0.3, 0.3) + it.velocity.z
+                        )
+                    )
+                )
+
+                hpMap[it] = it.health
+            }
```

Comment:
Instead of tracking each entity for health changes you could listen to Hurt Time Updates instead.

Reply:
Sorry, I'm not familiar with higher version of Fabric APIs